clang/docs/LanguageExtensions.rst

```diff
@@ -1826,6 +1827,12 @@ The following type trait primitives are supported by Clang. Those traits marked
   functionally equivalent to copying the underlying bytes and then dropping the
   source object on the floor. This is true of trivial types and types which
   were made trivially relocatable via the ``clang::trivial_abi`` attribute.
+* ``__builtin_is_cpp_trivially_relocatable`` (C++): Returns true if an object
+  is trivially relocatable, as defined by the C++26 standard [meta.unary.prop].
+  Note that the caller code should ensure that if the object is polymorphic,
+  the dynamic type is of the most derived type.
```

Comment:
Why do we introduce another builtin for this instead of fixing the existing one?

Reply:
I'd like a built in that has the standard behavior.
I think there is an interesting question as to whether the existing builtin should be modified to return true
if either a type is trivial_abi or relocatable by the standard. I think this would make sense

Reply:
Wouldn't it make a lot more sense to deprecate marking types as `trivial_abi` that aren't trivially relocatable? AFIAICT that's a requirement anyways, which is why it has been added in the first place. Then we can just say "__is_trivially_relocatable returns only true for types that are according to the standard" instead of introducing a new builtin just for the old builtin to turn out to be more useful in the end anyways.

Reply:
I would be happy doing that but I'm afraid it would break existing code.
Right now 

__is_cpp_trivially_relocatable => standard behavior
__is_trivially_relocatable => standard behavior + trivial_abi

Reply:
I'm not sure I understand what breakage you're afraid of. Sure, a deprecation warning would break stuff with `-Werror`, but that's nothing new. I'm not saying we should make `__is_trivially_relocatable` return false if only `[[clang::trivial_abi]]` is added. I'm saying we should fix the users code and warn (and eventually error) if they're not giving us the guarantee. I just don't see when we'd ever want to use `__is_cpp_trivially_relocatable` over `__is_trivially_relocatable`.

Reply:
`std::is_trivially_relocatable` should *not* return true for a type that is not relocatable per the standard.

This gives us a few options, none of them great
- Introduce a new trait so that we can model the standard behavior
- Modify the existing trait to ignore `[[clang::trivial_abi]]`  which _will_ break code
- Be not conforming

Having a warning does not solve the conformance issue.
I agree with you that we probably want
 - Restrict the set of types that can be  `[[clang::trivial_abi]]`  (ie warn if  `[[clang::trivial_abi]]`  is added to a non-relocatable type)
 - Converge to a single trait by deprecating the old trait.

Reply:
Corentin is right.  Any attempt we make to modify `__is_trivially_relocatable` is going to either be non-standard or break `[[clang::trivial_abi]]`.

SO, we'd want to choose which we'd want to do.  I very strongly don't believe we can be non-standard here, this is an important standardized feature, and `clang::trivial_abi` isn't any better.

I am ok with us modifying `clang::trivial_abi` over a number of releases(like 2-3 releases!) to make it be the same as the standards feature, but only if we can come to an agreement with the folks who are already using it.

SO in short, I think combining these builtins right now is untenable.  Long-term, it is  possible but only if we are willing/able to significantly break `clang::trivial_abi`.

Reply:
I made an issue for that https://github.com/llvm/llvm-project/issues/128725

Reply:
> `std::is_trivially_relocatable` should _not_ return true for a type that is not relocatable per the standard.

Why? It's our extension to define. Am I missing something that makes annotating a type `[[clang::trivial_abi]]` instead of `trivially_relocatable_if_eligible [[clang::trivial_abi]]` significantly different? AFAIK having a type marked `[[clang::trivial_abi]]` requires it to be trivially relocatable.

Reply:
> AFAIK having a type marked [[clang::trivial_abi]] requires it to be trivially relocatable.

It's not exactly the same semantics.

In particular, the standard traits allow for polymorphic types, any code that would do a memcpy without checking for that would lead to nasty UB.

How special members contribute to relocatability is also different, the standard requires non-deleted assignment.
That would also break code or be forever non-conforming and portable.